src/main/java/org/javarosa/core/model/instance/ExternalDataInstanceSource.java

```diff
@@ -1,64 +1,90 @@
-    }
+    @Nullable
+    private UUID storageReferenceId;
 
-    public ExternalDataInstanceSource(String instanceId, TreeElement root, String sourceUri, boolean useCaseTemplate) {
-        this(instanceId, root, sourceUri, ImmutableMultimap.of(), useCaseTemplate);
+    public final static String JR_REMOTE_REFERENCE = "jr://instance/remote";
+
+    public ExternalDataInstanceSource() {
     }
 
-    public ExternalDataInstanceSource(String instanceId, TreeElement root, String sourceUri, Multimap<String, String> requestData, boolean useCaseTemplate) {
+    private ExternalDataInstanceSource(String instanceId, TreeElement root, String reference, boolean useCaseTemplate,
+                                       String sourceUri, Multimap requestData, UUID storageReferenceId) {
+        if (sourceUri == null && storageReferenceId == null) {
+            throw new RuntimeException(getClass().getCanonicalName()
+                    + " must be initialised with one of sourceUri or storageReferenceId");
+        }
         this.instanceId = instanceId;
-        this.sourceUri = sourceUri;
-        this.requestData = requestData;
         this.root = root;
+        this.reference = reference;
         this.useCaseTemplate = useCaseTemplate;
+        this.sourceUri = sourceUri;
+        this.requestData = requestData;
+        this.storageReferenceId = storageReferenceId;
     }
 
     /**
      * Copy constructor
      */
-    public ExternalDataInstanceSource(ExternalDataInstanceSource externalDataInstanceSource)  {
+    public ExternalDataInstanceSource(ExternalDataInstanceSource externalDataInstanceSource) {
         this.instanceId = externalDataInstanceSource.instanceId;
-        this.sourceUri = externalDataInstanceSource.sourceUri;
         this.root = externalDataInstanceSource.root;
+        this.reference = externalDataInstanceSource.reference;
         this.useCaseTemplate = externalDataInstanceSource.useCaseTemplate();
+        this.sourceUri = externalDataInstanceSource.sourceUri;
         this.requestData = externalDataInstanceSource.requestData;
+        this.storageReferenceId = externalDataInstanceSource.storageReferenceId;
+    }
+
+    public static ExternalDataInstanceSource buildRemoteDataInstanceSource(
+            String instanceId, @Nullable TreeElement root,
+            boolean useCaseTemplate, String sourceUri,
+            Multimap requestData) {
+        return new ExternalDataInstanceSource(instanceId,root, JR_REMOTE_REFERENCE,
+                useCaseTemplate, sourceUri, requestData, null);
+    }
+
+
+    public static ExternalDataInstanceSource buildStorageBackedDataInstanceSource(
+            String instanceId, @Nullable TreeElement root,
+            String reference, boolean useCaseTemplate,
+            UUID storageReferenceId) {
+        return new ExternalDataInstanceSource(instanceId, root, reference,
+                useCaseTemplate, null, ArrayListMultimap.create(), storageReferenceId);
```

Comment:
Why change this from `ImmutableMultimap.of()` to `ArrayListMultimap.create()`? It should never be mutated should it?

Reply:
thanks for catching, it was unintentional.